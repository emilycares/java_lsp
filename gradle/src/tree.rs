use std::process::Command;

use serde::{Deserialize, Serialize};

#[derive(PartialEq, Debug, Serialize, Deserialize, Clone)]
pub struct Dependency<'a> {
    pub group_id: &'a str,
    pub artivact_id: &'a str,
    pub version: &'a str,
}

pub fn load<'a>() -> Option<Vec<Dependency<'a>>> {
    let log: String = get_cli_output()?;
    let out = parse_tree(log);
    Some(out)
}

fn get_cli_output() -> Option<String> {
    // ./gradlew dependencies --console --plain
    let output = Command::new("./gradlew")
        .arg("dependencies")
        .arg("--console")
        .arg("plain")
        // .arg("-b")
        .output()
        .ok()?;

    Some(String::from_utf8_lossy(&output.stdout).to_string())
}

fn parse_tree<'a>(inp: String) -> Vec<Dependency<'a>> {
    let mut out = vec![];

    let mut capture = false;

    for line in inp.lines() {
        if line.contains(" - ") && line.ends_with(".") {
            capture = true;
        }

        if line.starts_with("(c) - A dependency constraint") {
            break;
        }

        if capture {
            if line.contains(" - ")
                || line.is_empty()
                || !line.contains("-")
                || line.starts_with("No dependencies")
            {
            } else {
                let line = line
                    .replace("-", "")
                    .replace("\\", "")
                    .replace(" ", "")
                    .replace("+", "")
                    .replace("|", "")
                    .replace("(*)", "")
                    .replace("(n)", "")
                    .replace("(c)", "");
                let line: &'static str = Box::leak(line.into_boxed_str());
                let mut spl = line.splitn(3, ":");
                if let Some(group_id) = spl.next() {
                    if let Some(artivact_id) = spl.next() {
                        if let Some(version) = spl.next() {
                            out.push(Dependency {
                                group_id,
                                artivact_id,
                                version,
                            })
                        }
                    }
                }
            }
        }
    }

    out
}

#[cfg(test)]
mod tests {
    use crate::tree::parse_tree;

    #[test]
    fn parse_diagram() {
        let inp = include_str!("../tests/dependencies_report.txt");
        let out = parse_tree(inp.to_owned());
        insta::assert_yaml_snapshot!(out);
    }
}
