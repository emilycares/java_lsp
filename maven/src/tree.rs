use std::{process::Command, str::FromStr};

use nom::branch::alt;
use nom::{
    bytes::complete::{tag, take_until},
    multi::separated_list0,
    sequence::delimited,
    IResult,
};
use serde::{Deserialize, Serialize};

use crate::MavenError;

pub fn load<'a>() -> Result<Dependency<'a>, MavenError> {
    let log: String = get_cli_output()?;
    let cut: String = cut_output(log);
    // let mut output = File::create("/tmp/tree")?;
    // write!(output, "{}", cut)?;
    let input: &'static str = Box::leak(cut.into_boxed_str());
    let out = parser(input);
    match out {
        Ok(o) => Ok(o.1),
        Err(e) => Err(MavenError::TreeParseError(e)),
    }
}

fn get_cli_output() -> Result<String, MavenError> {
    // mvn dependency:tree -DoutputType=dot -b
    let output = Command::new("mvn")
        .arg("dependency:tree")
        .arg("-DoutputType=dot")
        // .arg("-b")
        .output()?;

    Ok(String::from_utf8_lossy(&output.stdout).to_string())
}

fn cut_output(inp: String) -> String {
    let mut out = String::new();

    let mut capture = false;

    for line in inp.lines() {
        if line.starts_with("[INFO] digraph") {
            capture = true;
        }

        if capture {
            out.push_str(line);
            out.push('\n');
        }

        if line.starts_with("[INFO]  }") {
            break;
        }
    }

    out
}

#[derive(PartialEq, Debug, Serialize, Deserialize)]
pub struct Dependency<'a> {
    #[serde(borrow)]
    pub base: Pom<'a>,
    pub deps: Vec<Pom<'a>>,
}

#[derive(PartialEq, Debug, Serialize, Deserialize)]
pub struct Pom<'a> {
    pub group_id: &'a str,
    pub artivact_id: &'a str,
    pub version: &'a str,
    pub scope: DependencyScope,
}

/// https://maven.apache.org/guides/introduction/introduction-to-dependency-mechanism.html#dependency-scope
#[derive(Default, PartialEq, Debug, Serialize, Deserialize)]
pub enum DependencyScope {
    #[default]
    Compile,
    Provided,
    /// No need to be indexed
    Runtime,
    /// Only considered in test
    Test,
    System,
    Import,
}

impl FromStr for DependencyScope {
    type Err = MavenError;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        match s {
            "compile" => Ok(Self::Compile),
            "provided" => Ok(Self::Provided),
            "runtime" => Ok(Self::Runtime),
            "test" => Ok(Self::Test),
            "system" => Ok(Self::System),
            "import" => Ok(Self::Import),
            _ => Err(MavenError::UnknownDependencyScope),
        }
    }
}

fn parse_pom(input: &str) -> IResult<&str, Pom> {
    let (input, group_id) = take_until(":")(input)?;
    let (input, _) = tag(":")(input)?;
    let (input, artivact_id) = take_until(":")(input)?;
    let (input, _) = tag(":")(input)?;
    let (input, _type) = take_until(":")(input)?;
    let (input, _) = tag(":")(input)?;
    let (input, version) = take_until("\"")(input)?;
    Ok((
        input,
        Pom {
            group_id,
            artivact_id,
            version,
            scope: DependencyScope::Test,
        },
    ))
}
fn parse_pom_b(input: &str) -> IResult<&str, Pom> {
    let (input, group_id) = take_until(":")(input)?;
    let (input, _) = tag(":")(input)?;
    let (input, artivact_id) = take_until(":")(input)?;
    let (input, _) = tag(":")(input)?;
    let (input, _type) = take_until(":")(input)?;
    let (input, _) = tag(":")(input)?;
    let (input, version) = take_until(":")(input)?;
    let (input, _) = tag(":")(input)?;
    let (input, scope) = take_until("\"")(input)?;
    Ok((
        input,
        Pom {
            group_id,
            artivact_id,
            version,
            scope: scope.parse().expect("UnknownDependencyScope"),
        },
    ))
}
fn parse_relation(input: &str) -> IResult<&str, Pom> {
    let (input, _) = take_until(" -> ")(input)?;
    let (input, _) = tag(" -> ")(input)?;
    let (input, out) = delimited(tag("\""), parse_pom_b, tag("\""))(input)?;

    Ok((input, out))
}

fn parser(input: &str) -> IResult<&str, Dependency> {
    let (input, _) = tag("[INFO] digraph ")(input)?;
    let (input, base) = delimited(tag("\""), parse_pom, tag("\""))(input)?;
    let (input, _) = alt((tag(" {\n[INFO]  "), tag(" { \n[INFO] \t")))(input)?;
    let (input, deps) = separated_list0(
        alt((tag(" ;\n[INFO]  "), tag(" ; \n[INFO] \t"))),
        parse_relation,
    )(input)?;

    let (input, _) = take_until("[INFO]")(input)?;
    let (input, _) = tag("[INFO]  }")(input)?;
    let (input, _) = take_until("\n")(input)?;
    let (input, _) = tag("\n")(input)?;
    Ok((input, Dependency { base, deps }))
}

#[cfg(test)]
mod tests {
    use crate::tree::{cut_output, parser};

    #[test]
    fn cut_basic() {
        let inp = include_str!("../tests/tverify.bacic.txt");

        let out = cut_output(inp.to_string());

        assert!(!out.contains("Building getting-started"));
        assert!(!out.contains("BUILD SUCCESS"));
    }

    #[test]
    fn parse_diagram() {
        let inp = include_str!("../tests/tverify.bacic.txt");
        let cut = cut_output(inp.to_string());
        let out = parser(&cut);
        let out = out.unwrap();
        insta::assert_yaml_snapshot!(out.1);
    }

    #[test]
    fn parse_diagram_with_tab() {
        let inp = include_str!("../tests/tverify-tap.bacic.txt");
        let out = parser(inp);
        let out = out.unwrap();
        insta::assert_yaml_snapshot!(out.1);
    }
}
