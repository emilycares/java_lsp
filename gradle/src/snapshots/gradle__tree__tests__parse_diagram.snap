---
source: gradle/src/tree.rs
expression: out
snapshot_kind: text
---
- group_id: com.github.toolarium
  artivact_id: toolariumenumconfiguration
  version: 1.2.0
- group_id: com.fasterxml.jackson.core
  artivact_id: jacksondatabind
  version: 2.17.1
- group_id: com.fasterxml.jackson.core
  artivact_id: jacksonannotations
  version: 2.17.1
- group_id: com.fasterxml.jackson
  artivact_id: jacksonbom
  version: 2.17.1
- group_id: com.fasterxml.jackson.core
  artivact_id: jacksonannotations
  version: 2.17.1
- group_id: com.fasterxml.jackson.core
  artivact_id: jacksoncore
  version: 2.17.1
- group_id: com.fasterxml.jackson.core
  artivact_id: jacksondatabind
  version: 2.17.1
- group_id: com.fasterxml.jackson.datatype
  artivact_id: jacksondatatypejsr310
  version: 2.17.1
- group_id: com.fasterxml.jackson.core
  artivact_id: jacksoncore
  version: 2.17.1
- group_id: com.fasterxml.jackson
  artivact_id: jacksonbom
  version: 2.17.1
- group_id: com.fasterxml.jackson
  artivact_id: jacksonbom
  version: 2.17.1
- group_id: com.fasterxml.jackson.core
  artivact_id: jacksoncore
  version: 2.17.1
- group_id: com.fasterxml.jackson.datatype
  artivact_id: jacksondatatypejsr310
  version: 2.17.1
- group_id: com.fasterxml.jackson.core
  artivact_id: jacksonannotations
  version: 2.17.1
- group_id: com.fasterxml.jackson.core
  artivact_id: jacksoncore
  version: 2.17.1
- group_id: com.fasterxml.jackson.core
  artivact_id: jacksondatabind
  version: 2.17.1
- group_id: com.fasterxml.jackson
  artivact_id: jacksonbom
  version: 2.17.1
- group_id: org.slf4j
  artivact_id: slf4japi
  version: 2.0.10
- group_id: com.puppycrawl.tools
  artivact_id: checkstyle
  version: 10.3.3
- group_id: info.picocli
  artivact_id: picocli
  version: 4.6.3
- group_id: org.antlr
  artivact_id: antlr4runtime
  version: 4.10.1
- group_id: commonsbeanutils
  artivact_id: commonsbeanutils
  version: 1.9.4
- group_id: commonscollections
  artivact_id: commonscollections
  version: 3.2.2
- group_id: com.google.guava
  artivact_id: guava
  version: 31.1jre
- group_id: com.google.guava
  artivact_id: failureaccess
  version: 1.0.1
- group_id: com.google.guava
  artivact_id: listenablefuture
  version: 9999.0emptytoavoidconflictwithguava
- group_id: com.google.code.findbugs
  artivact_id: jsr305
  version: 3.0.2
- group_id: org.checkerframework
  artivact_id: checkerqual
  version: 3.12.0
- group_id: com.google.errorprone
  artivact_id: error_prone_annotations
  version: 2.11.0
- group_id: com.google.j2objc
  artivact_id: j2objcannotations
  version: "1.3"
- group_id: org.reflections
  artivact_id: reflections
  version: 0.10.2
- group_id: org.javassist
  artivact_id: javassist
  version: 3.28.0GA
- group_id: com.google.code.findbugs
  artivact_id: jsr305
  version: 3.0.2
- group_id: net.sf.saxon
  artivact_id: SaxonHE
  version: "11.4"
- group_id: org.xmlresolver
  artivact_id: xmlresolver
  version: 4.4.3
- group_id: org.apache.httpcomponents.client5
  artivact_id: httpclient5
  version: 5.1.3
- group_id: org.apache.httpcomponents.core5
  artivact_id: httpcore5
  version: 5.1.3
- group_id: org.apache.httpcomponents.core5
  artivact_id: httpcore5h2
  version: 5.1.3
- group_id: org.apache.httpcomponents.core5
  artivact_id: httpcore5
  version: 5.1.3
- group_id: commonscodec
  artivact_id: commonscodec
  version: "1.15"
- group_id: org.apache.httpcomponents.core5
  artivact_id: httpcore5
  version: 5.1.3
- group_id: com.github.toolarium
  artivact_id: toolariumenumconfiguration
  version: 1.2.0
- group_id: org.slf4j
  artivact_id: slf4japi
  version: 2.0.13
- group_id: com.github.toolarium
  artivact_id: toolariumenumconfiguration
  version: 1.2.0
- group_id: org.slf4j
  artivact_id: slf4japi
  version: 2.0.13
- group_id: org.jacoco
  artivact_id: org.jacoco.agent
  version: 0.8.9
- group_id: org.jacoco
  artivact_id: org.jacoco.ant
  version: 0.8.9
- group_id: org.jacoco
  artivact_id: org.jacoco.core
  version: 0.8.9
- group_id: org.ow2.asm
  artivact_id: asm
  version: "9.5"
- group_id: org.ow2.asm
  artivact_id: asmcommons
  version: "9.5"
- group_id: org.ow2.asm
  artivact_id: asm
  version: "9.5"
- group_id: org.ow2.asm
  artivact_id: asmtree
  version: "9.5"
- group_id: org.ow2.asm
  artivact_id: asm
  version: "9.5"
- group_id: org.ow2.asm
  artivact_id: asmtree
  version: "9.5"
- group_id: org.jacoco
  artivact_id: org.jacoco.report
  version: 0.8.9
- group_id: org.jacoco
  artivact_id: org.jacoco.core
  version: 0.8.9
- group_id: org.jacoco
  artivact_id: org.jacoco.agent
  version: 0.8.9
- group_id: net.sf.jptools
  artivact_id: jptools
  version: 1.7.11
- group_id: net.sourceforge.jexcelapi
  artivact_id: jxl
  version: 2.6.12
- group_id: org.apache.poi
  artivact_id: poi
  version: "3.9"
- group_id: commonscodec
  artivact_id: commonscodec
  version: "1.5"
- group_id: org.apache.poi
  artivact_id: poiooxml
  version: "3.9"
- group_id: org.apache.poi
  artivact_id: poi
  version: "3.9"
- group_id: org.apache.poi
  artivact_id: poiooxmlschemas
  version: "3.9"
- group_id: org.apache.xmlbeans
  artivact_id: xmlbeans
  version: 2.3.0
- group_id: stax
  artivact_id: staxapi
  version: 1.0.1
- group_id: dom4j
  artivact_id: dom4j
  version: 1.6.1
- group_id: xmlapis
  artivact_id: xmlapis
  version: 1.0.b2
- group_id: org.slf4j
  artivact_id: slf4japi
  version: 2.0.13
- group_id: org.slf4j
  artivact_id: slf4japi
  version: 2.0.13
- group_id: org.junit.jupiter
  artivact_id: junitjupiterapi
  version: 5.7.1
- group_id: org.junit
  artivact_id: junitbom
  version: 5.7.1
- group_id: org.junit.jupiter
  artivact_id: junitjupiterapi
  version: 5.7.1
- group_id: org.junit.platform
  artivact_id: junitplatformcommons
  version: 1.7.1
- group_id: org.apiguardian
  artivact_id: apiguardianapi
  version: 1.1.0
- group_id: org.opentest4j
  artivact_id: opentest4j
  version: 1.2.0
- group_id: org.junit.platform
  artivact_id: junitplatformcommons
  version: 1.7.1
- group_id: org.junit
  artivact_id: junitbom
  version: 5.7.1
- group_id: org.apiguardian
  artivact_id: apiguardianapi
  version: 1.1.0
- group_id: com.github.toolarium
  artivact_id: toolariumenumconfiguration
  version: 1.2.0
- group_id: org.junit.jupiter
  artivact_id: junitjupiterapi
  version: 5.7.1
- group_id: com.github.toolarium
  artivact_id: toolariumenumconfiguration
  version: 1.2.0
- group_id: org.slf4j
  artivact_id: slf4japi
  version: 2.0.13
- group_id: org.junit.jupiter
  artivact_id: junitjupiterapi
  version: 5.7.1
- group_id: org.junit
  artivact_id: junitbom
  version: 5.7.1
- group_id: org.junit.jupiter
  artivact_id: junitjupiterapi
  version: 5.7.1
- group_id: org.junit.jupiter
  artivact_id: junitjupiterengine
  version: 5.7.1
- group_id: org.junit.platform
  artivact_id: junitplatformcommons
  version: 1.7.1
- group_id: org.junit.platform
  artivact_id: junitplatformengine
  version: 1.7.1
- group_id: org.junit.platform
  artivact_id: junitplatformlauncher
  version: 1.7.1
- group_id: org.apiguardian
  artivact_id: apiguardianapi
  version: 1.1.0
- group_id: org.opentest4j
  artivact_id: opentest4j
  version: 1.2.0
- group_id: org.junit.platform
  artivact_id: junitplatformcommons
  version: 1.7.1
- group_id: org.junit
  artivact_id: junitbom
  version: 5.7.1
- group_id: org.apiguardian
  artivact_id: apiguardianapi
  version: 1.1.0
- group_id: com.github.toolarium
  artivact_id: toolariumenumconfiguration
  version: 1.2.0
- group_id: com.fasterxml.jackson.core
  artivact_id: jacksondatabind
  version: 2.17.1
- group_id: com.fasterxml.jackson.core
  artivact_id: jacksonannotations
  version: 2.17.1
- group_id: com.fasterxml.jackson
  artivact_id: jacksonbom
  version: 2.17.1
- group_id: com.fasterxml.jackson.core
  artivact_id: jacksonannotations
  version: 2.17.1
- group_id: com.fasterxml.jackson.core
  artivact_id: jacksoncore
  version: 2.17.1
- group_id: com.fasterxml.jackson.core
  artivact_id: jacksondatabind
  version: 2.17.1
- group_id: com.fasterxml.jackson.datatype
  artivact_id: jacksondatatypejsr310
  version: 2.17.1
- group_id: com.fasterxml.jackson.core
  artivact_id: jacksoncore
  version: 2.17.1
- group_id: com.fasterxml.jackson
  artivact_id: jacksonbom
  version: 2.17.1
- group_id: com.fasterxml.jackson
  artivact_id: jacksonbom
  version: 2.17.1
- group_id: com.fasterxml.jackson.core
  artivact_id: jacksoncore
  version: 2.17.1
- group_id: com.fasterxml.jackson.datatype
  artivact_id: jacksondatatypejsr310
  version: 2.17.1
- group_id: com.fasterxml.jackson.core
  artivact_id: jacksonannotations
  version: 2.17.1
- group_id: com.fasterxml.jackson.core
  artivact_id: jacksoncore
  version: 2.17.1
- group_id: com.fasterxml.jackson.core
  artivact_id: jacksondatabind
  version: 2.17.1
- group_id: com.fasterxml.jackson
  artivact_id: jacksonbom
  version: 2.17.1
- group_id: org.slf4j
  artivact_id: slf4japi
  version: 2.0.10>2.0.13
- group_id: org.junit.jupiter
  artivact_id: junitjupiterengine
  version: 5.7.1
- group_id: org.junit
  artivact_id: junitbom
  version: 5.7.1
- group_id: org.apiguardian
  artivact_id: apiguardianapi
  version: 1.1.0
- group_id: org.junit.platform
  artivact_id: junitplatformengine
  version: 1.7.1
- group_id: org.junit
  artivact_id: junitbom
  version: 5.7.1
- group_id: org.apiguardian
  artivact_id: apiguardianapi
  version: 1.1.0
- group_id: org.opentest4j
  artivact_id: opentest4j
  version: 1.2.0
- group_id: org.junit.platform
  artivact_id: junitplatformcommons
  version: 1.7.1
- group_id: org.junit.jupiter
  artivact_id: junitjupiterapi
  version: 5.7.1
- group_id: org.junit
  artivact_id: junitbom
  version: 5.7.1
- group_id: org.apiguardian
  artivact_id: apiguardianapi
  version: 1.1.0
- group_id: org.junit.platform
  artivact_id: junitplatformengine
  version: 1.7.1
- group_id: ch.qos.logback
  artivact_id: logbackclassic
  version: 1.5.6
- group_id: ch.qos.logback
  artivact_id: logbackcore
  version: 1.5.6
- group_id: org.slf4j
  artivact_id: slf4japi
  version: 2.0.13
- group_id: org.junit.jupiter
  artivact_id: junitjupiterengine
  version: 5.7.1
- group_id: ch.qos.logback
  artivact_id: logbackclassic
  version: 1.5.6
